contrib/epee/include/net/abstract_tcp_server2.inl

```diff
@@ -81,860 +81,1049 @@ namespace net_utils
-  bool connection<t_protocol_handler>::do_send_chunk(byte_slice chunk)
+  template<typename T>
+  bool connection<T>::send(byte_slice_t message)
   {
-    TRY_ENTRY();
-    // Use safe_shared_from_this, because of this is public method and it can be called on the object being deleted
-    auto self = safe_shared_from_this();
-    if(!self)
-      return false;
-    if(m_was_shutdown)
+    lock_guard_t guard(state.lock);
+    if (state.status != status_t::RUNNING || state.socket.wait_handshake)
       return false;
-    double current_speed_up;
-    {
-		CRITICAL_REGION_LOCAL(m_throttle_speed_out_mutex);
-		m_throttle_speed_out.handle_trafic_exact(chunk.size());
-		current_speed_up = m_throttle_speed_out.get_current_speed();
-	}
-    context.m_current_speed_up = current_speed_up;
-    context.m_max_speed_up = std::max(context.m_max_speed_up, current_speed_up);
-
-    //_info("[sock " << socket().native_handle() << "] SEND " << cb);
-    context.m_last_send = time(NULL);
-    context.m_send_cnt += chunk.size();
-    //some data should be wrote to stream
-    //request complete
-    
-    // No sleeping here; sleeping is done once and for all in "handle_write"
-
-    m_send_que_lock.lock(); // *** critical ***
-    epee::misc_utils::auto_scope_leave_caller scope_exit_handler = epee::misc_utils::create_scope_leave_handler([&](){m_send_que_lock.unlock();});
-
-    long int retry=0;
-    const long int retry_limit = 5*4;
-    while (m_send_que.size() > ABSTRACT_SERVER_SEND_QUE_MAX_COUNT)
-    {
-        retry++;
-
-        /* if ( ::cryptonote::core::get_is_stopping() ) { // TODO re-add fast stop
-            _fact("ABORT queue wait due to stopping");
-            return false; // aborted
-        }*/
-
+    auto wait_consume = [this] {
+      auto random_delay = []{
         using engine = std::mt19937;
-
-        engine rng;
         std::random_device dev;
-        std::seed_seq::result_type rand[engine::state_size]{};  // Use complete bit space
-
+        std::seed_seq::result_type rand[
+          engine::state_size // Use complete bit space
+        ]{};
         std::generate_n(rand, engine::state_size, std::ref(dev));
         std::seed_seq seed(rand, rand + engine::state_size);
-        rng.seed(seed);
-
-        long int ms = 250 + (rng() % 50);
-        MDEBUG("Sleeping because QUEUE is FULL, in " << __FUNCTION__ << " for " << ms << " ms before packet_size="<<chunk.size()); // XXX debug sleep
-        m_send_que_lock.unlock();
-        boost::this_thread::sleep(boost::posix_time::milliseconds( ms ) );
-        m_send_que_lock.lock();
-        _dbg1("sleep for queue: " << ms);
-	if (m_was_shutdown)
-		return false;
-
-        if (retry > retry_limit) {
-            MWARNING("send que size is more than ABSTRACT_SERVER_SEND_QUE_MAX_COUNT(" << ABSTRACT_SERVER_SEND_QUE_MAX_COUNT << "), shutting down connection");
-            shutdown();
-            return false;
+        engine rng(seed);
+        return std::chrono::milliseconds(
```

Comment:
This is just plain stupid nonsense. You initialize RNG once and then use it, not seed it from std::random_device every time (~2500 bytes!!!). At this stage, you could've just used std::random_device directly. This implementation will deplete system's entropy very quickly.

Reply:
Yes, seeding the full state (2500 bytes) from std::random_device is totally ok and even good if you do it once. But this code must be rewritten.

Reply:
> This implementation will deplete system's entropy very quickly.

That's indeed bad for a crypto project, like Monero.

Reply:
It turns out this code has been there for a long time. Still, since this PR is almost a full rewrite, it's better to fix it like in #8365

Reply:
I agree. Large rewrites aren't tolerated here from anybody, and for valid reasons.

Plus we can't use double standards.